File: backend/app/api/v1/endpoints/performance.py
"""Measured performance: what the sources actually got right, counted from settled results.

Every number here is computed from stored score rows by `app.services.settlement`. Nothing is
estimated, and a source with nothing scored says "not measured yet" with the reason rather than
showing a zero. An accuracy figure is never published without the sample it was computed from, its
definition and the rule that settled it, and it is withheld entirely below the minimum sample.

No provider request is made by any endpoint in this file.
"""

from __future__ import annotations

from datetime import date, datetime, timedelta, timezone
from typing import Any, Dict, Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.core.deps import get_current_admin_user
from app.db.session import get_db
from app.models.users import User
from app.services import settlement as settlement_service

router = APIRouter()

MAX_WINDOW_DAYS = settlement_service.MAX_WINDOW_DAYS


def _validated_window(start: Optional[date], end: Optional[date]) -> Dict[str, date]:
    """Reject a window that cannot be answered honestly before doing any work."""
    today = datetime.now(timezone.utc).date()
    end_date = end or today
    start_date = start or (end_date - timedelta(days=settlement_service.DEFAULT_WINDOW_DAYS))
    if start_date > end_date:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST,
                            detail="start must not be after end")
    if (end_date - start_date).days + 1 > MAX_WINDOW_DAYS:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"the window may not exceed {MAX_WINDOW_DAYS} days; ask for a shorter range")
    return {"start": start_date, "end": end_date}


@router.get("/rules", summary="The settlement rules every score on this installation was computed with")
async def rules() -> Dict[str, Any]:
    """The full ruleset, published on its own so a figure can always be checked against it."""
    return settlement_service.settlement_rules()


@router.get("/sources", summary="Measured performance per source, counted from settled results")
async def sources(
    start: Optional[date] = Query(None, description="First kickoff date to include (UTC). Defaults to 90 days back."),
    end: Optional[date] = Query(None, description="Last kickoff date to include (UTC). Defaults to today."),
    db: Session = Depends(get_db),
) -> Dict[str, Any]:
    """Per source - each model provider and each expert - over a window of kickoffs.

    For every source: how many predictions were eligible, scored, still pending, void or refused,
    then per market the hit rate of the most likely outcome and a Brier score over the published
    probabilities. Both carry their sample size and their definition, and both are withheld with a
    stated reason below the minimum sample rather than shown from a handful of matches.
    """
    window = _validated_window(start, end)
    return settlement_service.measurement(db, start=window["start"], end=window["end"])


@router.post("/settle", summary="Score settled matches against the stored forecasts (admin)")
async def settle(
    start: Optional[date] = Query(None, description="First kickoff date to settle (UTC). Defaults to 90 days back."),
    end: Optional[date] = Query(None, description="Last kickoff date to settle (UTC). Defaults to today."),
    limit: Optional[int] = Query(None, ge=1, le=5000, description="Maximum matches to process in this run"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
) -> Dict[str, Any]:
    """Run settlement over a window. Idempotent: re-running changes nothing already scored.

    Reads only stored data - stored fixtures, stored results, stored forecast snapshots - and makes
    no provider request, so it is safe to run as often as wanted.
    """
    window = _validated_window(start, end)
    start_dt, end_dt = settlement_service.window_bounds(window["start"], window["end"])
    report = settlement_service.SettlementService(db).settle_range(start=start_dt, end=end_dt, limit=limit)
    report["window"] = {"start": window["start"].isoformat(), "end": window["end"].isoformat(),
                        "basis": "kickoff date in UTC, both ends included"}
    report["settled_by"] = str(current_user.id)
    return report

File: frontend/src/types/performance.ts
/**
 * The measured record: what each source actually got right, counted from settled results.
 *
 * Served by `GET /api/v1/performance/sources` (backend/app/api/v1/endpoints/performance.py, which
 * computes everything in `app/services/settlement.py`). Nothing in this payload is estimated: every
 * figure is arithmetic over stored score rows, and a figure the backend refused to compute arrives
 * as `null` beside the reason it was refused.
 *
 * THREE PROPERTIES OF THIS PAYLOAD THE INTERFACE MUST PRESERVE.
 *
 *  1. A figure never travels without its sample. `hit_rate` is meaningless without
 *     `hit_rate_sample`, and the backend withholds the rate entirely below `minimum_sample`
 *     rather than publishing one from a handful of matches. A withheld figure is `null` WITH a
 *     `*_unavailable_reason` — it is a sentence to print, never a zero, a dash or an empty bar.
 *  2. A figure never travels without its definition. `hit_rate_definition`, `brier_definition`
 *     and `rule` are the backend's own words for what was counted, and they are what makes the
 *     number checkable.
 *  3. A figure never travels without its window. `window` says which kickoffs were included and
 *     on what basis, so "62%" cannot be read as an all-time claim.
 *
 * Nothing here is a prediction, and nothing here may be presented as one: a measured record is a
 * statement about the past.
 */

/** The range of kickoffs a measurement covers, in the backend's own terms. */
export interface MeasuredWindow {
  /** First kickoff date included, ISO `YYYY-MM-DD`. */
  start: string;
  /** Last kickoff date included, ISO `YYYY-MM-DD`. */
  end: string;
  /** How the two dates are to be read, e.g. "kickoff date in UTC, both ends included". */
  basis: string;
}

/**
 * The settlement ruleset every figure on this installation was computed with.
 *
 * Published alongside the numbers on purpose: a hit rate is only checkable against the rule that
 * produced it. `markets` is an open map so a market added to the backend after this build shipped
 * still reaches the reader with its rule attached.
 */
export interface SettlementRules {
  version: string;
  basis: string;
  void: string;
  unsupplied_market: string;
  prematch_only: string;
  markets: Record<string, string>;
  probability_of_actual: string;
  brier: string;
  brier_baselines: Record<string, number>;
  hit_rate: string;
  minimum_sample: number;
  minimum_sample_rationale: string;
}

/**
 * One market's measured record for one source.
 *
 * `scored` is the denominator behind `hit_rate`: voids, pushes and markets the source never
 * published are excluded from it, which is why `scored + pushes + voids + not_scored` is the
 * eligible total rather than `scored` alone.
 */
export interface MeasuredMarket {
  /** Market key, e.g. `match_result`. Open: a market this build has never heard of still arrives. */
  market: string;
  /** How this market settles, in the backend's words. */
  rule: string;
  scored: number;
  hits: number;
  pushes: number;
  voids: number;
  not_scored: number;

  /** Share of `scored` whose most likely published outcome occurred. Null when withheld. */
  hit_rate: number | null;
  hit_rate_available: boolean;
  /** How many scored predictions the rate was (or would be) computed from. Always present. */
  hit_rate_sample: number;
  /** Why no rate is published, when none is. Print it; never substitute a zero. */
  hit_rate_unavailable_reason: string | null;
  hit_rate_definition: string;

  /** Mean squared error of the published probabilities. Lower is better. Null when withheld. */
  brier_score: number | null;
  brier_available: boolean;
  brier_unavailable_reason: string | null;
  brier_definition: string;
  /** What an uninformative forecast would score on this market, for comparison. May be null. */
  brier_baseline: number | null;
  brier_sample: number;

  /** Mean probability the source published for the outcome that occurred. Null when none applies. */
  mean_probability_of_actual: number | null;
  mean_probability_of_actual_definition: string;
  mean_probability_of_actual_sample: number;
}

/** One reason predictions were not scored, and how many it applied to. */
export interface MeasuredNotScoredReason {
  reason: string;
  count: number;
}

/**
 * One source over the window: a model provider, or an expert.
 *
 * `measured` is the gate. When it is false the source has NO headline figure and
 * `not_measured_reason` is the sentence to show instead — that path is the normal one on an
 * installation where settlement has not yet had results to score.
 */
export interface MeasuredSource {
  /** `model_provider` or `expert`; open, so a source kind added later still renders. */
  source_type: string;
  source_id: string;
  source_label: string;
  /** Predictions from this source that were in scope at all. */
  eligible: number;
  /** Predictions actually scored against a result. The sample behind everything else. */
  scored: number;
  /** Settled fixtures still waiting to be scored. */
  pending: number;
  /** Fixtures that were never played to a result: never a win, never a loss. */
  void: number;
  /** Eligible predictions the rules refused to score, with the reasons below. */
  not_scored: number;
  not_scored_reasons: MeasuredNotScoredReason[];
  markets: MeasuredMarket[];
  measured: boolean;
  /** Why this source has no measured record, when it has none. */
  not_measured_reason: string | null;
}

/** The whole `GET /api/v1/performance/sources` payload. */
export interface MeasuredPerformance {
  window: MeasuredWindow;
  /** Scored predictions required before any headline figure is published. */
  minimum_sample: number;
  minimum_sample_rationale: string;
  rules: SettlementRules;
  sources: MeasuredSource[];
  /** How many entries of `sources` have `measured: true`. */
  sources_measured: number;
  /** Why nothing at all is measured, when the source list is empty. */
  not_measured_reason: string | null;
  measured_at: string;
}
